File: exercism/ruby/clock/clock.rb
module BookKeeping
  VERSION = 2
end

class Clock

  attr_reader :hour, :minute

  def self.at(h, m = 0)
    min = m % 60
    hour = (h +  m / 60) % 24
    c = Clock.new(hour, min)
  end

  def initialize(h, m)
    @hour = h
    @minute = m
  end

  def to_s
    "%02d:%02d" % [@hour, @minute]
  end

  def +(delta)
    Clock.at(@hour, @minute + delta)
  end

  def -(delta)
    self + (-delta)
  end

  def ==(clock)
    @hour == clock.hour && @minute == clock.minute
  end
end

File: exercism/haskell/clock/src/Clock.hs
module Clock (fromHourMin, toString) where

import Text.Printf (printf)

data Time = Time Integer Integer
  deriving (Eq, Show)

fromHourMin :: Integer -> Integer -> Time
fromHourMin hour min = fromInteger (60 * hour + min)

toString :: Time -> String
toString (Time hour min) = printf "%02d:%02d" hour min

instance Num Time where
  (+) (Time h1 m1) (Time h2 m2) = fromHourMin (h1 + h2) (m1 + m2)
  (-) (Time h1 m1) (Time h2 m2) = fromHourMin (h1 - h2) (m1 - m2)
  negate = (-) (Time 24 00)
  fromInteger n = Time (mod (div n 60) 24) (mod n 60)
  (*) = const
  abs = id
  signum = id
